databroker_pack/_pack.py

```diff
@@ -306,15 +318,29 @@ def export_run(
                         root = root_map.get(doc["root"], doc["root"])
                         unique_id = root_hash_func(doc["root"])
                         if external is None:
-                            resource = doc.copy()
-                            resource["root"] = root
-                            files[(root, unique_id)].update(run.get_file_list(resource))
-                        # Replace root with a unique ID before serialization.
-                        # We are overriding the local variable name doc here
-                        # (yuck!) so that serializer(name, doc) below works on
-                        # all document types.
-                        doc = doc.copy()
-                        doc["root"] = unique_id
+                            if dry_run:
+                                root_in_document = doc["root"]
+                            else:
+                                root_in_document = root
+                            # - root_in_document is the 'root' actually in the
+                            # resource_document
+                            # - root may be different depending on the
+                            # source_catalog configuration, which can map the
+                            # recorded 'root' in the document to some other
+                            # location. This is where we should go looking for
+                            # the data if we plan to copy it.
+                            # - unique_id is unique to this (root, salt)
+                            # combination and used to place the data in a
+                            # unique location.
+                            key = (root_in_document, root, unique_id)
+                            files[key].update(run.get_file_list(doc))
+                        if not dry_run:
+                            # Replace root with a unique ID before serialization.
+                            # We are overriding the local variable name doc here
+                            # (yuck!) so that serializer(name, doc) below works on
```

Comment:
You do the same on L311 without editorial comment.

Reply:
True. As we've said elsewhere, it sure looks like `DocumentRouter` needs a callback registry or a streamz dependency.